### PR TITLE
feat: add verifyWithIndividualChecks

### PR DESCRIPTION
**Pull request by sebastianquek**
Returns an array of promises, each promise corresponds to a verification check.
The last promise resolves to the overall validity based on all the checks.

This allows us to work on the promises independently from the other verification checks.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -3,6 +3,37 @@ import { verifyHash } from "./hash/hash";
 import { verifyIssued } from "./issued/verify";
 import { verifyRevoked } from "./revoked/verify";
 import { documentToSmartContracts } from "./common/smartContract/documentToSmartContracts";
+import { OpenAttestationContract } from "./types";
+
+type ResolveType<T> = T extends (...args: any[]) => Promise<infer U> ? U : T;
```
Comment (Nebulis) [inline src/index.ts]: I think this one deserve a comment to explain what we are trying (comment + example maybe)

File: src/verify.integration.test.ts [hunk 2]
```diff
@@ -152,3 +152,174 @@ describe("verify(integration)", () => {
     );
   });
 });
+
+describe("verifyWithIndividualChecks(integration)", () => {
+  it("returns false if document is invalid", async () => {
```
Comment (Nebulis) [inline src/verify.integration.test.ts]: I think `document is invalid` is not clear enough on the intent, From the expectation:
- document has not been issued
- hash is invalid

What are we testing in that case ?

File: src/verify.test.ts [hunk 3]
```diff
@@ -61,3 +63,56 @@ describe("verify", () => {
     });
   });
 });
+
+describe("verifyWithIndividualChecks", () => {
```
Comment (Nebulis) [inline src/verify.test.ts]: what does this test bring compare to integration test ?

File: src/verify.integration.test.ts [hunk 4]
```diff
@@ -152,3 +152,174 @@ describe("verify(integration)", () => {
...
+            address: "0x48399Fb88bcD031C556F53e93F690EEC07963Af3",
+            revoked: false
+          }
+        ]
+      },
+      valid: true
+    });
+  });
+
+  it("returns false if Ropsten token is invalid", async () => {
```
Comment (Nebulis) [inline src/verify.integration.test.ts]: I dont understand the meaning of this ? what are we testing :thinking:

Comment (Nebulis) [inline src/verify.integration.test.ts]: Same as above, the test description is not describing enough what we are testing
- document has not been issued
- document has been revoked

What are we testing ?

File: src/index.ts (hunk 1 above)
Comment (sebastianquek) [reply src/index.ts]: OK, I've added some comments about this

File: src/verify.test.ts (hunk 3 above)
Comment (sebastianquek) [reply src/verify.test.ts]: The integration test ensures the checks work on an actual document with the correct network specified (also ensures the various checks work properly together).

The unit test, on the other hand, mocks what's returned by the various checks and checks whether the returned values are correct given these mocks. Also, we can mimic the way the various promises resolve at different times.

How should we better spilt the test concerns?

Comment (sebastianquek) [reply src/verify.test.ts]: As discussed, I've removed the unit tests that don't value add much.

**Pull request merged**